### DISCUSSION TITLE
Unexpected behaviour difference between overload and type union, when mapping interpreted as iterable

### DISCUSSION THREAD

**Discussion by tommilligan** [microsoft/pyright#7836] [Q&A] [answered]
I have an unexpected behaviour on latest pyright (v1.1.361) - I'm not sure if this is a bug in pyright, or pyright pointing out some subtle complexity that mypy doesn't catch. Would appreciate another pair of eyes on it before I raise a bug report!

### Hypothesis

Under some circumstances, when given a `dict[K, V]` for a parameter `Mapping[K, V] | Iterable[T]`, pyright preferentially treats `dict` as an `Iterable`, causing a type error of the form `K is incompatible with T` (as a dictionary is treated as an iterable of keys).

However, this ignores the fact that `dict[K, V]` is compatible with `Mapping[K, V]`.

### Question

Is this intended behaviour, and if it is not, is there a path to resolving it?

### Minimal reproduction

We define some generic type variables `K` and `V`, and some function that returns the generic type `dict[K, V]`:

```python
from collections.abc import Iterable, Mapping
from typing import TypeVar, overload

K = TypeVar("K")
V = TypeVar("V")


def into_dict(mapping: Mapping[K, V]) -> dict[K, V]:
    """Return a mapping as a dict, using generic types."""
    return dict(mapping)
```

We separately define a function that accepts key-value pairs. It accepts the data either in a `Mapping`, or in an `Iterable` of `tuple`s, which we define using `overload`. For simplicity, this function is empty:

```python
@overload
def drop_overload(value: Mapping[K, V]) -> None: ...

@overload
def drop_overload(value: Iterable[tuple[K, V]]) -> None: ...

def drop_overload(value) -> None:
    """Do nothing with a value."""
```

We can call this function as expected, with the output of the function returning `dict[K, V]` with the specific type `dict[str, str]`:

```python
drop_overload(into_dict({"example-key": "example-value"}))  # no type error
```

We can simplify the above overloads, by defining a new function using a type union for the input argument:

```python
def drop_union(value: Mapping[K, V] | Iterable[tuple[K, V]]) -> None:
    """Do nothing with a value."""
```

However, this doesn't behave in the same way. Specifically, when the output of the function with generic signature `dict[K, V]` is used as input, we get an error.

This error does not happen if a constant value or variable is passed in directly:

```python
drop_union(into_dict({"example-key": "example-value"}))  # type error
# mvp.py:42:23 - error: Argument of type "dict[int, str]" cannot be assigned to parameter "mapping" of type "Mapping[K@into_dict, V@into_dict]" in function "into_dict"
#     "Literal['example-key']" is incompatible with "tuple[K@drop_union, V@drop_union]" (reportArgumentType)

# Note that there is no type error, if the output of `into_dict` is assigned to a variable first.
data = into_dict({"example-key": "example-value"})
drop_union(data)  # fine
```

A full reproduction is attached that can be tested with `pyright mvp.txt`: [mvp.txt](https://github.com/microsoft/pyright/files/15192892/mvp.txt)

**Comment by tommilligan** [comment]
After some different keyword searching in the discussions, I found https://github.com/microsoft/pyright/discussions/7292#discussioncomment-8513592 which seems to suggest that this is a partially known/expected behaviour, i.e. matching a `dict[K, V]` to an `Iterable[tuple[K, V]]` type is correct and it would be a behaviour change to alter this.

However, I'm still not clear on why we have different behaviour for:
- Overloads vs Type Unions
- Direct function output vs variables/constant values

In my opinion, this still points out a bug (or maybe just some documentation clarification on how these heuristics are applied?), which I'd be happy to contribute towards if someone could point me to the right are of the codebase to add a failing unit test in.

**Reply by erictraut** [reply, marked as answer]
The error you're seeing is due to pyright's [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). 

Here's a simplified version of your code:

```python
from collections.abc import Iterable, Mapping

def foo[A, B](v: Mapping[A, B]) -> dict[A, B]: ...
def bar[C, D](v: Mapping[C, D] | Iterable[tuple[C, D]]) -> None: ...

bar(foo({"a": "b"}))
```

When pyright evaluates the expression `foo({"a": "b"})`, it takes into account the type expected by the corresponding parameter for function `bar`. In this case, that type is `Mapping[K, V] | Iterable[tuple[K, V]]`. The return type of `foo` is `dict[K, V]`. Before evaluating the call to `foo`, pyright needs to place constraints on type variables `A` and `B`. In this case, the constraints that should be applied are ambiguous because both `dict[A, B]` is compatible with both `Mapping[C, D]` and `Iterable[tuple[C, D]]`. One set of constraints is `A = C, B = D`. Another set of constraints is `A = tuple[C, D]`. Pyright uses a heuristic to pick between these. In this case, it happens to choose the wrong one (`Iterable` rather than `Mapping`), and the evaluation fails.

The problem doesn't occur with an overload because the logic does not need to choose between two potentially valid types in that case.

Is this a bug? Arguably yes. I could introduce a better (more complicated) heuristic. That said, all heuristics are fallible in some cases.

**Reply by tommilligan** [reply]
> One set of constraints is A = C, B = D. Another set of constraints is A = tuple[C, D]. Pyright uses a heuristic to pick between these. In this case, it happens to choose the wrong one (Iterable rather than Mapping), and the evaluation fails.

Right, this makes sense, thanks for the clarification. Could you point out to me where this heuristic is applied? I agree that changing/making it more complex is probably not what we want, but I'd be interested to see the existing approach.

> The problem doesn't occur with an overload because the logic does not need to choose between two potentially valid types in that case.

I'm still not sure why this is the case - could you elaborate on the difference of a type union vs. multiple overloads? In the example I posted where two overloads are provided, there is surely still a choice between two potentially valid types for the parameter at the callsite? Is it that this choice happens later, when specific types are already known?

To rephrase, why does applying the "incorrect" constraint `A = tuple[C, D]` not happen in the overload case, where that is a valid call? Could you point me to where overload signatures are matched in the code?

**Reply by erictraut** [reply]
In the overload case, the overload signatures are applied one at a time until one of them succeeds. If you look at your overload signatures, none of them involves a union of `Mapping` and `Iterable`.

**Comment by tommilligan** [comment]
Thanks for your help on this - I feel like I have a much better understanding of the problem now.

I agree that changing the heuristic for type assignment is not worth it and likely to get messy, given the workaround of an overload exists.

I see there are already some notes on heuristics in `mypy-comparison.md` - I'll send a PR with an addition summarising the above discussion.

---

For my own reference, my understanding is:

- Here is where type specialization in union types occurs: https://github.com/microsoft/pyright/blob/7be85982d6fd609f9b26d5ff05b6b72626f1dd65/packages/pyright-internal/src/analyzer/typeEvaluator.ts#L24563
- Here is where multiple overloads are evaluated in turn: https://github.com/microsoft/pyright/blob/7be85982d6fd609f9b26d5ff05b6b72626f1dd65/packages/pyright-internal/src/analyzer/typeEvaluator.ts#L24013

They are both called in the core `assignType` function: https://github.com/microsoft/pyright/blob/7be85982d6fd609f9b26d5ff05b6b72626f1dd65/packages/pyright-internal/src/analyzer/typeEvaluator.ts#L23142

**Reply by erictraut** [reply]
Thanks for offering, but a PR isn't necessary. I don't think this topic belongs in the documentation. Mypy has similar heuristics and behaviors in its bidirectional type inference. There are numerous open issues in the mypy issue tracker relating to this.
